README.md
# dither

[![go reportcard](https://goreportcard.com/badge/github.com/makeworld-the-better-one/dither)](https://goreportcard.com/report/github.com/makeworld-the-better-one/dither)
[![Go Reference](https://pkg.go.dev/badge/github.com/makeworld-the-better-one/dither/v2.svg)](https://pkg.go.dev/github.com/makeworld-the-better-one/dither/v2)

dither is a library for dithering images in Go. It has many dithering algorithms built-in, and allows you to specify your own. Correctness is a top priority, as well as performance. It is designed to work well on its own, but also implements interfaces from the standard library, so that it can be integrated easily in a wide variety of situtations.

This library is uniquely correct from a math and quality perspective. It linearizes the image, and color comparisons are done with human luminance perception in mind (channel weighting). Few-to-no other libraries do this.

It supports images that make use of the alpha channel, AKA transparency.

*Make sure to set your browser zoom to 100% to view these images properly*

| **Original** | **Dithering Algorithm** |
| :-: | :-: |
| | Floyd-Steinberg (black and white palette) |
| [![](images/examples/david.png)](https://raw.githubusercontent.com/makeworld-the-better-one/dither/master/images/examples/david.png) | [![](images/examples/floyd-steinberg_david.png)](https://raw.githubusercontent.com/makeworld-the-better-one/dither/master/images/examples/floyd-steinberg_david.png) |
| | Floyd-Steinberg (red, green, yellow, black) |
| [![](images/examples/peppers_small.png)](https://raw.githubusercontent.com/makeworld-the-better-one/dither/master/images/examples/peppers_small.png) | [![](images/examples/floyd-steinberg_peppers.png)](https://raw.githubusercontent.com/makeworld-the-better-one/dither/master/images/examples/floyd-steinberg_peppers.png)

## Types of dithering supported

- Random noise (in grayscale and RGB)
- **Ordered Dithering**
  - Bayer matrix of any size (as long as dimensions are powers of two)
  - Clustered-dot - many different preprogrammed matrices
  - Some unusual horizontal or vertical line matrices
  - Yours?
    - Using `PixelMapperFromMatrix`, this library can dither using any matrix
    - If you need more freedom, `PixelMapper` can be used to implement any method of dithering that affects each pixel individually
- **Error diffusion dithering**
  - Simple 2D
  - Floyd-Steinberg, False Floyd-Steinberg
  - Jarvis-Judice-Ninke
  - Atkinson
  - Stucki
  - Burkes
  - Sierra/Sierra3, Sierra2, Sierra2-4A/Sierra-Lite
  - [Steven Pigeon](https://hbfs.wordpress.com/2013/12/31/dithering/)
  - Yours? Custom error diffusion matrices can be used by the library.

More methods of dithering are being worked on, such as Riemersma, Yuliluoma, and blue noise.

## Install

In your project, run

```
go get github.com/makeworld-the-better-one/dither/v2@latest
```

You can import it as `"github.com/makeworld-the-better-one/dither/v2"` and use it as `dither`.

## Usage

Here's a simple example using Floyd-Steinberg dithering.

```go
img := // Get image.Image from somewhere

// These are the colors we want in our output image
palette := []color.Color{
    color.Black,
    color.White,
    // You can put any colors you want
}

// Create ditherer
d := dither.NewDitherer(palette)
d.Matrix = dither.FloydSteinberg

// Dither the image, attempting to modify the existing image
// If it can't then a dithered copy will be returned.
img = d.Dither(img)

// Now use img - save it as PNG, display it on the screen, etc
```

If you always want to dither a copy of the image, you can use `DitherCopy` instead.

Here's how you create a `Ditherer` that does Bayer dithering. Note how `d.Mapper` is used instead of `d.Matrix`.

```go
d := dither.NewDitherer(palette)
d.Mapper = dither.Bayer(8, 8, 1.0) // 8x8 Bayer matrix at 100% strength
```

Here's how you create a `Ditherer` that does clustered-dot dithering - dithering with a predefined matrix.

```go
d := dither.NewDitherer(palette)
d.Mapper = dither.PixelMapperFromMatrix(dither.ClusteredDotDiagonal8x8)
```

See the [docs](https://pkg.go.dev/github.com/makeworld-the-better-one/dither/v2) for more.


### More Examples

Sometimes you can't dither using the above code. These examples show how you can use this library in those situations.

- [Output to a static GIF](examples/gif_image/main.go)
- [Output to an animated GIF](examples/gif_animation/main.go)

If you're interested in what specific algorithms look like, you can check out the tests [output](images/output/) folder.

## Performance

Operations that only affect each pixel individually are parallelized, using `runtime.GOMAXPROCS(0)` which defaults to the number of CPUs. This applies to any `PixelMapper` (aka `Ditherer.Mapper`) but not to an `ErrorDiffusionMatrix` (aka `Ditherer.Matrix`), as the latter is inherently sequential.


## Scaling images

A dithered output image will only look right at 100% size. As you scale *down*, the image will immediately get darker, and strange grid-like artifacts will appear, known as a [moiré pattern](https://en.wikipedia.org/wiki/Moir%C3%A9_pattern). This is due to how dithered images work, and is not something this library can fix.

The best thing to do is to scale the *input* image to the *exact* size you want before using this library. But sometimes you want to scale the image up after dithering, to make the dithering effect more obvious for aesthetic purposes.

So for scaling the dithered output image *up* (above 100%), that will only look fine if you use **nearest-neighbor scaling** - the kind of scaling that produces pixelated results. Otherwise the dither pixel values will be blurred and averaged, which will mess things up. And even once you're using that, it will still produce moiré patterns, unless you're scaling by a multiple of the original dimensions. **So when scaling up, you should be scaling by 2x or 3x, rather than a non-integer like 1.34x.**

## Encoding output

Dithered images require that their pixel values be stored exactly. This means they must be encoded to a lossless format. PNG is almost always the best choice, as it is widely supported and takes up the least space. GIF is also acceptable, as long as the palette is 256 colors or less. The GIF format is also useful if you are dithering an animation. APNG is more efficient for animation, but has no Go stdlib support, and less support in non-browser environments.

The WebP format also works for both static images and animation, but it must be a lossless WebP, not a lossy one.

## What method should I use?

Generally, using Floyd-Steinberg serpentine dithering will produce the best results. The code would be:

```go
d := dither.NewDitherer(yourPalette)
d.Matrix = dither.FloydSteinberg
d.Serpentine = true
```

Playing with the strength of the matrix might also be useful. The example above is at full strength, but sometimes that's too noisy. The code for 80% strength looks like this:

```go
d.Matrix = dither.ErrorDiffusionStrength(dither.FloydSteinberg, 0.8)
```

The main reason for using any other dithering algorithm would be

- **Aesthetics** - dithering can be a cool image effect, and different methods will look different
- **Speed** - error diffusion dithering is sequential and therefore single-threaded. But ordered dithering, like using `Bayer`, will use all available CPUs, which is much faster.

## How do I get the palette?

Sometimes the palette isn't an option, as it might determined by the hardware. Many e-ink screens can only display black and white for example, and so your palette is chosen for you.

But in most cases you have all the colors available, and so you have to pick the ones that represent your image best. This is called [color quantization](https://en.wikipedia.org/wiki/Color_quantization).

I might end up writing another library that implements some common algorithms for this, like median cut. But there are some libraries that exist already. [joshdk/quantize](https://github.com/joshdk/quantize) looks like the best one, although there is also [this one](https://pkg.go.dev/github.com/soniakeys/quant/median).

## Tips

Some general tips for working with the library.

Any returned `PixelMappers` should be cached and re-used. There is no point in regenerating them, it just wastes resources.

If the palette is grayscale, the input image should be converted to grayscale first to get accurate results.

All the `[][]uint` matrices are supposed to be applied with `PixelMapperFromMatrix`.


## Images with transparency

Images with transparency are only supported in v2.2.0 and after.

This library does not dither in the alpha channel or support transparent palettes. Instead it just keeps track of the alpha channel, and the dithered image returned will always have the exact same alpha values for each pixel. This allows for dithering of images with transparent parts.

Dithering images with semi-transparent pixels will also work, but is not as useful, because the output image will *appear* to have colors that are not in the palette, due to whatever background image you use.


## Projects using `dither`

- [didder](https://github.com/makeworld-the-better-one/didder) - a powerful CLI dithering tool
- [wasm-palette-converter](https://github.com/CrispyBaguette/wasm-palette-converter) - dithering images to a color palette, in the browser
- Your project? Build something fun, show how cool dithering can be! Some ideas / things I'd love to see:
  - A client-side web app for general-purpose dithering using WASM
  - A GUI desktop application


## Similar libraries
- `image/draw` has a `draw.Drawer` called `FloydSteinberg`
- https://github.com/brouxco/dithering
- https://github.com/esimov/dithergo

The largest problem with all of these libraries is that they don't linearize the image colors before dithering, which produces incorrect results. They also only support error diffusion dithering.


## License

This library is under the Mozilla Public License 2.0. Similar to the LGPL, this means you can use this library in your project, even if it's proprietary. But any changes you make to the library's code must be released publicly. Crucially, this license allows for statically linking this library.

See [LICENSE](LICENSE) for details, and my [blog post](https://www.makeworld.space/2021/01/lgpl_go.html) on why you should use the MPL over the LGPL for Go code.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.4.0] - 2023-12-20
### Changed
- Increased error diffusion dithering speed by ~50%
- Reduced error diffusion dithering memory usage by ~70% (#13)

### Fixed
- Docs: the input does still need to be converted to grayscale for grayscale palettes, actually (#7)

## [2.3.0] - 2022-12-20
### Changed
- When comparing colors, each channel is weighted according to human luminance perception (didder#14)

### Fixed
- Corrected Burkes matrix (#10)
- Palette order no longer affects output (#9)
- ~~Grayscale palettes don't require the input image be converted to grayscale beforehand (#7)~~ (incorrect, see #7)

## [2.2.0] - 2021-05-09
### Added
- Support for images with transparency (#8)


## [2.1.1] - 2021-04-30
### Changed
- Update Bayer strength recommendations for color images


## [2.1.0] - 2021-04-29
### Added
- JSON tags for `OrdereredDitherMatrix`

### Changed
- `Dither` never returns `nil`, making code simpler

### Fixed
- Bug where paletted images would never be detected as needing to be copied in `Dither`
- Palette is actually fully copied when needed, before the colors were shared with passed slice
  - `NewDitherer`
  - `Ditherer.GetColorPalette`
  - `DitherPaletted` and `DitherPalettedConfig`
  - `GetColorModel`


## [2.0.0] - 2021-02-13
### Added
- Added `ErrorDiffusionStrength` to set the strength of error diffusion dithering (#4)
- `RoundClamp` function for making your own `PixelMappers` that round correctly

### Changed
- All linear RGB values are represented using `uint16` instead of `uint8` now, because 8-bits is not enough to accurately hold a linearized value. This is a breaking change, hence the new major version.

### Fixed
- Rounding is no longer biased, because ties are rounded to the nearest even number


## [1.0.0] - 2021-02-11
Initial release.


ordered_ditherers.go
package dither

// This file contains matrices I've found from around the Internet. They can
// be used with PixelMapperFromMatrix.

// OrderedDitherMatrix is used to hold a matrix used for ordered dithering. This
// is useful if you find a matrix somewhere and would like to try it out. You can
// create this struct and then give it to PixelMapperFromMatrix.
//
// The matrix must be rectangular - each slice inside the first one must be the same
// length.
//
// Max is the value all the matrix values will be divided by. Usually this is the
// product of the dimensions of the matrix (x*y), or the greatest value in the matrix
// plus one. For diagonal matrices or other matrices with repeated values, it is the
// latter.
//
// Leaving Max as 0 will cause a panic.
//
// Matrix values should almost always range from 0 to Max-1. If the matrix you found
// ranges from 1 to Max, just subtract 1 from every value when programming it.
type OrderedDitherMatrix struct {
	Matrix [][]uint `json:"matrix"`
	Max    uint     `json:"max"`
}

// ClusteredDot4x4 comes from http://caca.zoy.org/study/part2.html
//
// It is not diagonal, so the dots form a grid.
var ClusteredDot4x4 = OrderedDitherMatrix{
	Matrix: [][]uint{
		{12, 5, 6, 13},
		{4, 0, 1, 7},
		{11, 3, 2, 8},
		{15, 10, 9, 14},
	},
	Max: 16,
}

// ClusteredDotDiagonal8x8 comes from http://caca.zoy.org/study/part2.html
//
// They say it "mimics the halftoning techniques used by newspapers". It is called
// "Diagonal" because the resulting dot pattern is at a 45 degree angle.
var ClusteredDotDiagonal8x8 = OrderedDitherMatrix{
	Matrix: [][]uint{
		{24, 10, 12, 26, 35, 47, 49, 37},
		{8, 0, 2, 14, 45, 59, 61, 51},
		{22, 6, 4, 16, 43, 57, 63, 53},
		{30, 20, 18, 28, 33, 41, 55, 39},
		{34, 46, 48, 36, 25, 11, 13, 27},
		{44, 58, 60, 50, 9, 1, 3, 15},
		{42, 56, 62, 52, 23, 7, 5, 17},
		{32, 40, 54, 38, 31, 21, 19, 29},
	},
	Max: 64,
}

// Vertical5x3 comes from http://caca.zoy.org/study/part2.html
//
// They say it "creates artistic vertical line artifacts".
var Vertical5x3 = OrderedDitherMatrix{
	Matrix: [][]uint{
		{9, 3, 0, 6, 12},
		{10, 4, 1, 7, 13},
		{11, 5, 2, 8, 14},
	},
	Max: 15,
}

// Horizontal3x5 is my rotated version of Vertical5x3.
var Horizontal3x5 = OrderedDitherMatrix{
	Matrix: [][]uint{
		{9, 10, 11},
		{3, 4, 5},
		{0, 1, 2},
		{6, 7, 8},
		{12, 13, 14},
	},
	Max: 15,
}

// ClusteredDotDiagonal6x6 comes from Figure 5.4 of the book Digital Halftoning by
// Robert Ulichney. In the book it's called "M = 3". It can represent "19 levels
// of gray". Its dimensions are 6x6, but as a diagonal matrix it is 7x7. It is called
// "Diagonal" because the resulting dot pattern is at a 45 degree angle.
var ClusteredDotDiagonal6x6 = OrderedDitherMatrix{
	Matrix: [][]uint{
		// 1 was subtracted from all original values so the range of values begins
		// at 0. Also the diagonal matrix was converted into a rectangular one.
		{8, 6, 7, 9, 11, 10},
		{5, 0, 1, 12, 17, 16},
		{4, 3, 2, 13, 14, 15},
		{9, 11, 10, 8, 6, 8},
		{12, 17, 16, 5, 0, 1},
		{13, 14, 15, 4, 3, 2},
	},
	Max: 18, // (x*y)/2 because it's diagonal
}

// ClusteredDotDiagonal8x8_2 comes from Figure 5.4 of the book Digital Halftoning by
// Robert Ulichney. In the book it's called "M = 4". It can represent "33 levels
// of gray". Its dimensionsare 8x8, but as a diagonal matrix it is 9x9. It is called
// "Diagonal" because the resulting dot pattern is at a 45 degree angle.
//
// It is almost identical to ClusteredDotDiagonal8x8, but worse because it can
// represent fewer gray levels. There is not much point in using it.
var ClusteredDotDiagonal8x8_2 = OrderedDitherMatrix{
	Matrix: [][]uint{
		// 1 was subtracted from all original values so the range of values begins
		// at 0. Also the diagonal matrix was converted into a rectangular one.
		{13, 11, 12, 15, 18, 20, 19, 16},
		{4, 3, 2, 9, 27, 28, 29, 22},
		{5, 0, 1, 10, 26, 31, 30, 21},
		{8, 6, 7, 14, 23, 25, 24, 17},
		{18, 20, 19, 16, 13, 11, 12, 15},
		{27, 28, 29, 22, 4, 3, 2, 9},
		{26, 31, 30, 21, 5, 0, 1, 10},
		{23, 25, 24, 17, 8, 6, 7, 14},
	},
	Max: 32, // (x*y)/2 because it's diagonal
}

// ClusteredDotDiagonal16x16 comes from Figure 5.4 of the book Digital Halftoning by
// Robert Ulichney. In the book it's called "M = 8". It can represent "129 levels
// of gray". Its dimensions are 16x16, but as a diagonal matrix it is 17x17. It is
// called "Diagonal" because the resulting dot pattern is at a 45 degree angle.
var ClusteredDotDiagonal16x16 = OrderedDitherMatrix{
	Matrix: [][]uint{
		// 1 was subtracted from all original values so the range of values begins
		// at 0.
		{63, 58, 50, 40, 41, 51, 59, 60, 64, 69, 77, 87, 86, 76, 68, 67},
		{57, 33, 27, 18, 19, 28, 34, 52, 70, 94, 100, 109, 108, 99, 93, 75},
		{49, 26, 13, 11, 12, 15, 29, 44, 78, 101, 114, 116, 115, 112, 98, 83},
		{39, 17, 4, 3, 2, 9, 20, 42, 87, 110, 123, 124, 125, 118, 107, 85},
		{38, 16, 5, 0, 1, 10, 21, 43, 89, 111, 122, 127, 126, 117, 106, 84},
		{48, 25, 8, 6, 7, 14, 30, 45, 79, 102, 119, 121, 120, 113, 97, 82},
		{56, 32, 24, 23, 22, 31, 35, 53, 71, 95, 103, 104, 105, 96, 92, 74},
		{62, 55, 47, 37, 36, 46, 54, 61, 65, 72, 80, 90, 91, 81, 73, 66},
		{64, 69, 77, 87, 86, 76, 68, 67, 63, 58, 50, 40, 41, 51, 59, 60},
		{70, 94, 100, 109, 108, 99, 93, 75, 57, 33, 27, 18, 19, 28, 34, 52},
		{78, 101, 114, 116, 115, 112, 98, 83, 49, 26, 13, 11, 12, 15, 29, 44},
		{87, 110, 123, 124, 125, 118, 107, 85, 39, 17, 4, 3, 2, 9, 20, 42},
		{89, 111, 122, 127, 126, 117, 106, 84, 38, 16, 5, 0, 1, 10, 21, 43},
		{79, 102, 119, 121, 120, 113, 97, 82, 48, 25, 8, 6, 7, 14, 30, 45},
		{71, 95, 103, 104, 105, 96, 92, 74, 56, 32, 24, 23, 22, 31, 35, 53},
		{65, 72, 80, 90, 91, 81, 73, 66, 62, 55, 47, 37, 36, 46, 54, 61},
	},
	Max: 128, // (x*y)/2 because it's diagonal
}

// ClusteredDot6x6 comes from Figure 5.9 of the book Digital Halftoning by
// Robert Ulichney. It can represent "37 levels of gray". It is not diagonal.
var ClusteredDot6x6 = OrderedDitherMatrix{
	Matrix: [][]uint{
		// 1 was subtracted from all original values so the range of values begins
		// at 0.
		{34, 29, 17, 21, 30, 35},
		{28, 14, 9, 16, 20, 31},
		{13, 8, 4, 5, 15, 19},
		{12, 3, 0, 1, 10, 18},
		{27, 7, 2, 6, 23, 24},
		{33, 26, 11, 22, 25, 32},
	},
	Max: 36,
}

// ClusteredDotSpiral5x5 comes from Figure 5.13 of the book Digital Halftoning by
// Robert Ulichney. It can represent "26 levels of gray". Its dimensions are 5x5.
//
// Instead of alternating dark and light dots like the other clustered-dot
// matrices, the dark parts grow to fill the area.
var ClusteredDotSpiral5x5 = OrderedDitherMatrix{
	Matrix: [][]uint{
		// 1 was subtracted from all original values so the range of values begins
		// at 0.
		{20, 21, 22, 23, 24},
		{19, 6, 7, 8, 9},
		{18, 5, 0, 1, 10},
		{17, 4, 3, 2, 11},
		{16, 15, 14, 13, 12},
	},
	Max: 25,
}

// ClusteredDotHorizontalLine comes from Figure 5.13 of the book Digital Halftoning by
// Robert Ulichney. It can represent "37 levels of gray". Its dimensions are 6x6.
//
// It "clusters pixels about horizontal lines".
var ClusteredDotHorizontalLine = OrderedDitherMatrix{
	Matrix: [][]uint{
		// 1 was subtracted from all original values so the range of values begins
		// at 0.
		{35, 33, 31, 30, 32, 34},
		{23, 21, 19, 18, 20, 22},
		{11, 9, 7, 6, 8, 10},
		{5, 3, 1, 0, 2, 4},
		{17, 15, 13, 12, 14, 16},
		{29, 27, 25, 24, 26, 28},
	},
	Max: 36,
}

// ClusteredDotVerticalLine is my rotated version of ClusteredDotHorizontalLine.
var ClusteredDotVerticalLine = OrderedDitherMatrix{
	Matrix: [][]uint{
		{35, 23, 11, 5, 17, 29},
		{33, 21, 9, 3, 15, 27},
		{31, 19, 7, 1, 13, 25},
		{30, 18, 6, 0, 12, 24},
		{32, 20, 8, 2, 14, 26},
		{34, 22, 10, 4, 16, 28},
	},
	Max: 36,
}

// ClusteredDot8x8 comes from Figure 1.5 of the book Modern Digital Halftoning,
// Second Edition, by Daniel L. Lau and Gonzalo R. Arce. It is like
// ClusteredDotDiagonal8x8, but is not diagonal. It can represent "65 gray-levels".
var ClusteredDot8x8 = OrderedDitherMatrix{
	Matrix: [][]uint{
		// For some reason the values in the book were in the range of 0-64
		// instead of 0-63. I changed the 64 value to 63, so that pure black
		// didn't end up with occasional white dots.
		{3, 9, 17, 27, 25, 15, 7, 1},
		{11, 29, 38, 46, 44, 36, 23, 5},
		{19, 40, 52, 58, 56, 50, 34, 13},
		{31, 48, 60, 63, 62, 54, 42, 21},
		{30, 47, 59, 63, 61, 53, 41, 20},
		{18, 39, 51, 57, 55, 49, 33, 12},
		{10, 28, 37, 45, 43, 35, 22, 4},
		{2, 8, 16, 26, 24, 14, 6, 0},
	},
	Max: 64,
}

// ClusteredDot6x6_2 comes from https://archive.is/71e9G. On the webpage it is
// called "central white point" while ClusteredDot6x6 is called "clustered dots".
//
// It is nearly identical to ClusteredDot6x6.
var ClusteredDot6x6_2 = OrderedDitherMatrix{
	Matrix: [][]uint{
		{34, 25, 21, 17, 29, 33},
		{30, 13, 9, 5, 12, 24},
		{18, 6, 1, 0, 8, 20},
		{22, 10, 2, 3, 4, 16},
		{26, 14, 7, 11, 15, 28},
		{35, 31, 19, 23, 27, 32},
	},
	Max: 36,
}

// ClusteredDot6x6_3 comes from https://archive.is/71e9G. On the webpage it is
// called "balanced centered point".
//
// It is nearly identical to ClusteredDot6x6.
var ClusteredDot6x6_3 = OrderedDitherMatrix{
	Matrix: [][]uint{
		{30, 22, 16, 21, 33, 35},
		{24, 11, 7, 9, 26, 28},
		{13, 5, 0, 2, 14, 19},
		{15, 3, 1, 4, 12, 18},
		{27, 8, 6, 10, 25, 29},
		{32, 20, 17, 23, 31, 34},
	},
	Max: 36,
}

// ClusteredDotDiagonal8x8_3 comes from https://archive.is/71e9G. On the webpage
// it is called "diagonal ordered matrix with balanced centered points".
//
// It is almost identical to ClusteredDotDiagonal8x8, but worse because it can
// represent fewer gray levels. There is not much point in using it.
//
// It is called "Diagonal" because the resulting dot pattern is at a 45 degree angle.
var ClusteredDotDiagonal8x8_3 = OrderedDitherMatrix{
	Matrix: [][]uint{
		// Yes, values are repeated. This is because the diamond matrix is formed
		// by combining two square matrices in a grid. See the linked page for more.
		{13, 9, 5, 12, 18, 22, 26, 19},
		{6, 1, 0, 8, 25, 30, 31, 23},
		{10, 2, 3, 4, 21, 29, 28, 27},
		{14, 7, 11, 15, 17, 24, 20, 16},
		{18, 22, 26, 19, 13, 9, 5, 12},
		{25, 30, 31, 23, 6, 1, 0, 8},
		{21, 29, 28, 27, 10, 2, 3, 4},
		{17, 24, 20, 16, 14, 7, 11, 15},
	},
	Max: 32,
}
